agent.go

```diff
@@ -0,0 +1,346 @@
+			if err != nil {
+				logger.Error("Error executing action", "error", err)
+				return err
+			}
+
+			// Record observation
+			observation := fmt.Sprintf("Observation from %s:\n%s", reActResp.Action.Name, output)
+			a.Trace(ctx, "system", observation)
+			a.Messages = append(a.Messages, Message{Role: "system", Content: observation})
+		}
+
+		a.CurrentIteration++
+	}
+
+	// Handle max iterations reached
+	logger.Info("Max iterations reached", "iterations", a.CurrentIteration)
+	fmt.Printf("\nSorry, Couldn't complete the task after %d attempts.\n", a.MaxIterations)
+	return fmt.Errorf("max iterations reached")
+}
+
+// executeAction handles the execution of a single action
+func (a *Agent) executeAction(ctx context.Context, action *Action, workDir string) (string, error) {
+	logger := loggerFromContext(ctx)
+
+	switch action.Name {
+	case "kubectl":
+		output, err := kubectlRunner(action.Input, a.Kubeconfig, workDir)
+		if err != nil {
+			return fmt.Sprintf("Error executing kubectl command: %v", err), err
+		}
+		return output, nil
+	case "cat":
+		output, err := bashRunner(action.Input, workDir, a.Kubeconfig)
+		if err != nil {
+			return fmt.Sprintf("Error executing cat command: %v", err), err
+		}
+		return output, nil
+	case "bash":
+		output, err := bashRunner(action.Input, workDir, a.Kubeconfig)
+		if err != nil {
+			return fmt.Sprintf("Error executing bash command: %v", err), err
+		}
+		return output, nil
+	default:
+		a.Trace(ctx, "system", fmt.Sprintf("Error: Tool %s not found", action.Name))
+		logger.Info("Unknown action: ", "action", action.Name)
+		return "", fmt.Errorf("unknown action: %s", action.Name)
+	}
+}
+
+func (a *Agent) AskLLM(ctx context.Context) (*ReActResponse, error) {
+	logger := loggerFromContext(ctx)
+	logger.Info("Asking LLM...")
+
+	data := Data{
+		Query:       a.Query,
+		PastQueries: a.PastQueries,
+		History:     a.History(),
+		Tools:       "kubectl, gcrane, bash",
+	}
+
+	prompt, err := a.generateFromTemplate(data)
+	if err != nil {
+		fmt.Println("Error generating from template:", err)
+		return nil, err
+	}
+
+	logger.Info("Asking LLM", "prompt", prompt)
+	WriteToFile(a.promptFilePath, prompt)
+	WriteToFile(a.promptFilePath, "\n========\n")
+
+	logger.Info("Thinking...", "prompt", prompt)
+
+	return a.ContentGenerator.GenerateContent(ctx, a.Model, prompt)
+}
+
+func sanitizeToolInput(input string) string {
+	return strings.TrimSpace(input)
+}
+
+func (a *Agent) Trace(ctx context.Context, role, content string) error {
+	logger := loggerFromContext(ctx)
+	logger.Info("Tracing...")
+	msg := Message{
+		Role:    role,
+		Content: content,
+	}
+	if role != "system" {
+		a.Messages = append(a.Messages, msg)
+	}
+	if a.tracePath == "" {
+		return nil
+	}
+	if err := WriteToFile(a.tracePath, fmt.Sprintf("%s: %s\n", role, content)); err != nil {
+		logger.Error("Error writing to trace file", "error", err)
+		// tracing is secondary to the main logic, so ignore the failures.
```

Comment:
We might want to pull out the tracing into its own package, or just use logging / structured logging / spans